### PR TITLE
compute broadcast join costs with pipeline dop

### PR DESCRIPTION
**Pull request by Youngwb**
#2201

### CODE REVIEW

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java [hunk 1]
```diff
@@ -256,24 +257,28 @@ public CostEstimate visitPhysicalDistribution(PhysicalDistributionOperator node,
...
+                            sessionVariable.getPipelineDop() : 1;
                     // beNum is the number of right table should broadcast, now use alive backends
                     int beNum = Math.max(1, Catalog.getCurrentSystemInfo().getBackendIds(true).size());
                     result = CostEstimate
                             .of(statistics.getOutputSize(outputColumns) *
                                             Catalog.getCurrentSystemInfo().getBackendIds(true).size(),
-                                    statistics.getOutputSize(outputColumns) * beNum * parallelExecInstanceNum,
-                                    statistics.getOutputSize(outputColumns) * beNum * parallelExecInstanceNum);
+                                    statistics.getOutputSize(outputColumns) * beNum * parallelExecInstanceNum *
+                                            pipelineDop,
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:279]: pipelineDop is 0 means that the query executed in pipeline engine use the half number of cpu cores. so here suggest to recifty pipelineDop here to a reasonable value when it is 0. 
i.e.:
`beNum * parallelExecInstanceNum * pipelineDop<=0?32:pipelineDop`

In future, the number of cores of BEs should be perceived by FE.

**Review by satanson — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java (hunk 1 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:283]: Done

**Review by satanson — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java [hunk 2]
```diff
@@ -256,24 +257,32 @@ public CostEstimate visitPhysicalDistribution(PhysicalDistributionOperator node,
...
-                            ConnectContext.get().getSessionVariable().getMaxExecMemByte()) {
+                    if (statistics.getOutputSize(outputColumns) > sessionVariable.getMaxExecMemByte()) {
                         return CostEstimate.infinite();
                     }
                     int parallelExecInstanceNum = Math.max(1, getParallelExecInstanceNum(context));
+                    int pipelineDop = 1;
+                    if (sessionVariable.isEnablePipelineEngine()) {
+                        // pipelineDop is 0 means that the query executed in pipeline engine use the half number of cpu
+                        // cores, we use 32 as an estimate value
+                        pipelineDop = sessionVariable.getPipelineDop() > 0 ? sessionVariable.getPipelineDop() : 32;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:275]: Add a TODO here.

32 isn't reasonable here.  `half number of cpu cores` could be any value. we should get this information from BE heartbeat.

Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:276]: done

**Review by kangkaisen — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Pull request merged**